server/WFLY-10083_Hot_Class_Replacement.adoc

```diff
@@ -0,0 +1,132 @@
+= Title
+:author:            Stuart Douglas
+:email:             sdouglas@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+The idea of this JIRA is to add hot class replacement support using Fakereplace to
+WildFly, so that hot replacement is supported 'out of the box'.
+
+This will be a gradual process that will involve several iterations. This document
+contains the requirements for the initial support.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-10083[WFLY-10083]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-975[EAP7-975]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Affected Projects or Components
+
+ * Wildfly
+ * Wildfly Core
+ * Undertow
+ * Weld
+ * Hibernate
+
+=== Other Interested Projects
+
+ * This may affect all components that work on user provided classes.
+
+== Requirements
+
+=== Hard Requirements
+
+The Fakereplace library will be shipped with Wildfly, and an option added to the startup scripts to run with the agent
+enabled. This will not be enabled by default, but will require the user to start wildfly with a specific option
+(tentatively `--developer-mode`).
+
+In general there are three specific parts to this support:
+
+* Figuring out which classes need to be replaced, and providing them to the Agent
+* Updating the bytecode
+* Informing any frameworks that use these classes, so they can rebuild their internal data structures to reflect the
+changes
+
+===== Class change notification
+
+Hot replacement will only be supported for exploded deployments. There are two different ways that classes can be
+replaced:
+
+ * By replacing the classes inside the exploded deployment. The hot replacement support will install an Undertow handler
+ inside the deployment chain, that will check for changed classes when a request arrives (in much the same manner as the
+ old Seam 2 hot replacement filter).
+ * Using a debugger over the JDWP. In this case the class bytes will be updated on the filesystem
+ to keep the class file in sync with what is loaded into the VM. This means that if the
+ developer restarts the server or deployment the class change will be persistent.
+
+The main difficulty in supporting non-exploded deployments is that it is generally not possible to update the jar file
+with the new classes, so if the deployment is restarted it will have the old versions of the classes. It may be possible
+to address this in future via various strategies, such as queuing the updates and applying them on redeploy/restart.
+
+==== Replacing the Bytecode
+
+The Fakereplace agent will be responsible for replacing the bytecode of any changed classes. This operation is not
+guaranteed to succeed, as some changes are too invasive to be handled by the agent (e.g. changing the superclass or
```

Comment:
Is it possible to add some sort of notification (log, event) when this operation does not succeed?